rust/src/module_expressions.rs

```diff
@@ -37,22 +42,29 @@ impl FromStr for ModuleExpression {
     type Err = GrimpError;
 
     fn from_str(expression: &str) -> Result<Self, Self::Err> {
-        if !MODULE_EXPRESSION_PATTERN.is_match(expression) {
-            return Err(GrimpError::InvalidModuleExpression(expression.to_owned()));
-        }
+        // Handle optional fragments by expanding into all possible combinations
+        let expanded_expressions = Self::expand_optional_fragments(expression);
 
-        for (part, next_part) in expression.split(".").tuple_windows() {
-            match (part, next_part) {
-                ("*", "**") | ("**", "*") | ("**", "**") => {
-                    return Err(GrimpError::InvalidModuleExpression(expression.to_owned()));
+        let mut patterns = Vec::new();
+        for expanded in &expanded_expressions {
+            // Validate each expanded expression
```

Comment:
The idea here is that after expanding the module expression, all of the expanded expressions should be valid by the old rules. E.g. `a.b[.**]` expands to `a.b` and `a.b.**`, both of which are valid expressions.